### PR TITLE
fix(cli): resolve repository root when building from a monorepo subdirectory

### PR DESCRIPTION
**Pull request by jeffsee55**
## Problem

When a project is linked **in place** — `apps/api/.vercel/project.json` — and `vc build` is run from that directory (directly, or via `turbo run` inside the app), the build treats the linked subdirectory as the repository root. Because the project's dependencies are typically hoisted to the monorepo root *above* it, this breaks builds in several ways that all share one root cause:

- **Redundant `rootDirectory` (very common misconfig):** a link at `apps/api` whose project also has `rootDirectory: "apps/api"` double-appends → `apps/api/apps/api`, failing with `ENOENT … /apps/api/apps/api/.next/package.json` (NEXT-4944).
- **`--standalone` runtime failure:** the pnpm resolution symlink (`apps/api/node_modules/hono` → `../../node_modules/.pnpm/.../hono`) is skipped because its target points outside the subdirectory, so the deployed function fails at runtime with `Cannot find module 'hono'` even though the dependency's files were packaged (PIPE-6621).
- **Wrong trace root:** builders trace from the subdirectory, so Next.js sets an incorrect `outputFileTracingRoot`/`turbopack.root` (Turbopack errors; Webpack `.nft.json` omits hoisted deps like `sharp`/`next`).

It shouldn't matter which directory the command is run from — but today it does.

## Fix (opt-in via `VERCEL_RESOLVE_ROOT_DIRECTORY=1`)

Resolve a per-directory link the same way a repository-level (`repo.json`) link already resolves: detect the repository root (workspace markers — `pnpm-workspace.yaml`, `package.json` `workspaces`, `lerna.json`, `rush.json` — then git, then the directory itself) and express the project as its path **relative to that root**. The build re-anchors to the repo root, so it's correctly rooted regardless of the invocation directory.

The link's **physical location is authoritative**: the project lives where the `.vercel` folder is. A `rootDirectory` setting is advisory — nullish or a redundant restatement is absorbed silently; anything else is ignored (location wins) with a clear advisory. A per-directory link never builds anywhere other than where it sits, so the common misconfig "just works."

For `--standalone`, package-manager symlinks are now **preserved** rather than skipped. Once the build is correctly rooted the symlink targets already resolve inside the function, so keeping the link is all that's required for `require('hono')` to work at runtime.

Behavior is unchanged when the flag is not set.

## Verified against the real reproduction

`vercel-support/standalone-repro` (`apps/api` linked in place):
- `rootDirectory: null` → resolves `repoRoot=<repo>`, builds `apps/api`, `hono` resolves from an isolated copy of the function (mirrors `/var/task`).
- `rootDirectory: "apps/api"` → no more `apps/api/apps/api`; builds cleanly.
- `rootDirectory: "apps/web"` (mismatch) → advisory printed, builds `apps/api` anyway.

## Testing

- **`repo-root.test.ts`**: root detection (pnpm/npm/yarn/lerna/rush, highest-match, fallbacks) and `resolvePerDirectoryLinkRoot` (null / redundant / mismatch / deeper-as-mismatch / at-root).
- **`monorepo.test.ts`**: integration tests — Hono and Next.js `--standalone` from a subdirectory (resolve the dependency from an isolated function), a non-standalone subdirectory build, and configs #3/#4. Verified to **fail without the fix** (config #4 reproduced the `apps/api/apps/api` crash; standalone reproduced `Cannot find module`).
- Existing legacy-mode standalone tests remain green (flag off).

Fixes PIPE-6621. Addresses NEXT-4944.

### CODE REVIEW

File: packages/cli/src/util/build/write-build-result.ts [hunk 1]
```diff
@@ -972,25 +972,25 @@ export function filesWithoutFsRefs(
...
     if (file.type === 'FileFsRef') {
       if (!filePathMap) filePathMap = {};
-      if (standalone && sharedDest) {
-        // pnpm and other package managers create symlinks in node_modules that
-        // point outside the app directory (e.g. ../../node_modules/.pnpm/...).
-        // These targets are rejected during prebuilt deploys, so skip them.
-        // The traced dependency files are included at their logical paths.
+      if (standalone && sharedDest && repoRootResolved) {
+        // Keep symlinks so bare imports resolve at runtime.
+        out[normalizePath(path)] = file;
```
Comment (ijjk) [inline packages/cli/src/util/build/write-build-result.ts]: I think this direct-write branch can race in `download()`. With `VERCEL_RESOLVE_ROOT_DIRECTORY=1`, standalone builds now keep `FileFsRef`s for pnpm-style symlinks in `files`, so `download(files, dest)` writes both `node_modules/<pkg>` and descendants like `node_modules/<pkg>/...` concurrently.

`download()` only skips descendants when the parent symlink is not “external”; pnpm symlinks usually point to `../../node_modules/.pnpm/...`, so the descendants are not skipped. If a descendant wins the race, its `mkdirp()` creates `node_modules/<pkg>` as a real directory. Then the symlink write hits `EEXIST`, attempts `readlink()` on that directory, and fails with `EINVAL`.

I was able to reproduce this locally with `@vercel/build-utils` `download()` using a pnpm-style symlink plus child file entries. Could we make this path deterministic, e.g. create symlinks before descendants, reclassify these targets as internal relative to the re-rooted function layout, or avoid sending the parent symlink and descendants through parallel `download()` together?

Comment (jeffsee55) [reply packages/cli/src/util/build/write-build-result.ts]: Good catch, confirmed and fixed in d50ddb9. In the resolved-root branch I now precompute the symlink entries and skip any `FileFsRef` whose key is nested under one, so only the symlink itself is materialized (its target still holds the real files in the `.pnpm` store). That keeps the symlink and its descendants from racing through `download()` together. Added a regression test in `write-build-result.test.ts` that asserts the descendant is dropped while the symlink and the real store file are kept (it fails without the fix).

**Review by ijjk — APPROVED** [review]
One compatibility edge case worth calling out before this graduates beyond the flag: this makes the physical location of a per-directory link authoritative and treats any non-matching `rootDirectory` as advisory/ignored.

That fixes the redundant `apps/api` setting case, but it changes behavior for less common setups where the link lives at one directory and intentionally points deeper or elsewhere, e.g. `apps/api/.vercel/project.json` with `rootDirectory: "server"` expecting `apps/api/server`, or a sibling path. With this change and `VERCEL_RESOLVE_ROOT_DIRECTORY=1`, those would build from `apps/api` instead.

If that behavior is intentional, the warning is probably enough while this is flag-gated. Before making this default, we should either document the migration path clearly or consider supporting relative subpaths under the link directory separately from redundant repo-relative values.

**Review by tknickman — APPROVED** [review]
